ably/realtime_conn.go

```diff
@@ -393,18 +393,32 @@ func (c *Connection) connectWith(arg connArgs) (result, error) {
 	}
 	u.RawQuery = query.Encode()
 	proto := c.opts.protocol()
+
+	if c.State() == ConnectionStateClosed { // RTN12d - if connection is closed by client, don't try to reconnect
+		return nopResult, nil
+	}
+
+	// if err is nil, raw connection with server is successful
 	conn, err := c.dial(proto, u)
 	if err != nil {
 		return nil, err
+	} else {
+		if ctxCancel != nil {
+			ctxCancel() // break reconnection loop, eventloop will be responsible for next reconnection iterations
```

Comment:
Maybe I'm missing something but I don't see what is this accomplishing.

`ctxCancel()` is called here when `err == nil`, to break the reconnection loop. But the reconnection loop is broken anyway, because when `connectWith`  returns a nil error, we `return` and stop the loop, in both cases:

```go
res, err := c.connectWith(arg, parentCtxCancel)
if err == nil {
    return res, nil
}
```

```go
res, err := c.connectWith(arg, parentCtxCancel)
if err != nil {
	if recoverable(err) {
		c.setState(ConnectionStateSuspended, err, retryIn)
		continue
	}
	return nil, c.setState(ConnectionStateFailed, err, 0)
}
return res, nil
```

Reply:
Okay, it seems it stops the loop either way, maybe I missed on that one.